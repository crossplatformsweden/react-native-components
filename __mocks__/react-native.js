module.exports = ({
  StyleSheet: {
    create: (styles) => styles,
  },
  Platform: {
    OS: 'ios',
  },
  Dimensions: {
    get: () => ({
      width: 1,
      height: 1,
    }),
  },
  View: 'View',
  Text: 'Text',
  Button: 'Button',
  ScrollView: 'View',
  Linking: { openURL: (url) => {console.log('Lingking.openURL', url)}},
  AsyncStorage: {
    setItem: () => new Promise((resolve) => resolve(), null),
    getItem: (key) =>
      new Promise((resolve) => resolve(`{"key": "${key}"}`), null),
    removeItem: (key, callback) => {
      if (callback) {
        callback();
      }
      return new Promise((resolve) => resolve(), null);
    },
  },
  Animated: {
    View: 'View',
    Value: jest.fn()
  },
  PanResponder: {
    create: () => 'View'
  },
  PanResponderInstance: 'View',
  TouchableOpacity: 'View',
  FlatList: 'View',
});
