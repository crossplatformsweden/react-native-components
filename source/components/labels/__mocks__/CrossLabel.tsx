export const CrossLabel = 'View';

export default CrossLabel;
