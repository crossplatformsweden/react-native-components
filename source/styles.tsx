import { Dimensions, Platform, StyleSheet } from 'react-native';
import { DefaultTheme } from 'react-native-paper';
const { width, height } = Dimensions.get('window');

const screenWidth = width < height ? width : height;

export enum Colors {
  CrossLightBlue = '#057dc2',
  CrossDarkBlue = '#00539b',
  CrossYellow = '#fede00',
  CrossBlack = '#2f2f2f',
  CrossWhite = 'white',
  CancelButton = '#ff0000',
  NextButton = '#349b34',
  BackButton = '#3180e0',
}

/**
 * Size if icons in tab bars
 * @type int
 * @public
 */
export const TabIconSize: number = Platform.OS === 'android' ? 22 : 22;

const IS_IOS = Platform.OS === 'ios';
const { width: viewportWidth, height: viewportHeight } = Dimensions.get(
  'window'
);

function wp(percentage: number) {
  const value = (percentage * viewportWidth) / 100;
  return Math.round(value);
}

export const slideHeight = viewportHeight * 0.15;
const itemWidth = wp(70);
const itemHorizontalMargin = wp(2);

export const sliderWidth = viewportWidth;
export const slideWidth = itemWidth + itemHorizontalMargin * 2;

const entryBorderRadius = 8;

/**
 * Theme appiled to React-Native-Paper
 *
 * https://callstack.github.io/react-native-paper/theming.html
 */
export const Theme = {
  ...DefaultTheme,
  roundness: 4,
  dark: false,
  colors: {
    ...DefaultTheme.colors,
    primary: Colors.CrossLightBlue,
    accent: Colors.CrossYellow,
    background: 'white',
    placeholder: Colors.CrossDarkBlue,
    backdrop: 'white',
    disabled: Colors.CrossDarkBlue,
    text: Colors.CrossBlack,
  },
};

export const CarouselStyles = StyleSheet.create({
  imageContainer: {
    flex: 1,
    marginBottom: IS_IOS ? 0 : -1, // Prevent a random Android rendering issue
    backgroundColor: 'white',
    borderTopLeftRadius: entryBorderRadius,
    borderTopRightRadius: entryBorderRadius,
  },
  image: {
    ...StyleSheet.absoluteFillObject,
    resizeMode: 'cover',
    borderRadius: IS_IOS ? entryBorderRadius : 0,
    borderTopLeftRadius: entryBorderRadius,
    borderTopRightRadius: entryBorderRadius,
  },
  // image's border radius is buggy on iOS; let's hack it!
  radiusMask: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    height: entryBorderRadius,
    backgroundColor: 'white',
  },
});

/**
 * Main theme for the app
 */
export const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
  },
  surface: {
    padding: 8,
    height: 80,
    width: 80,
    alignItems: 'center',
    justifyContent: 'center',
    elevation: 4,
  },
  buttonList: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    alignItems: 'flex-start',
  },
  listButton: {
    height: 40,
    flex: 1,
  },
  listItem: {
    margin: 0,
    padding: 0,
  },
  button: {
    marginRight: 10,
    height: 45,
  },
  paddingDefault: {
    paddingTop: 20,
    paddingLeft: 10,
    paddingRight: 10,
  },
  label: {
    color: 'gray',
    fontSize: 14,
  },
  link: {
    color: 'blue',
    fontWeight: 'bold',
  },
  tabIcon: {
    // Icons in tab bar won't show without explicit width / height
    width: 25,
    height: TabIconSize,
    color: 'white',
  },
  absoluteCentered: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
  absoluteTop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  absoluteTopRight: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    alignItems: 'flex-end',
    justifyContent: 'flex-start',
  },
  absoluteTopLeft: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  absoluteBottom: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  absoluteBottomRight: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'flex-end',
    justifyContent: 'flex-end',
  },
  absoluteBottomLeft: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'flex-start',
    justifyContent: 'flex-end',
  },
  centerVerticalAndHorizontal: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  alignCenterJustifyStart: {
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  alignStartJustifyStart: {
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  alignEndJustifyStart: {
    alignItems: 'flex-end',
    justifyContent: 'flex-start',
  },
  tabBar: {
    backgroundColor: Colors.CrossLightBlue,
  },
  tabBarTitle: {
    alignSelf: 'center',
  },
  tabBarLabel: {
    fontSize: 9,
  },
  rowContentTopCenter: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  rowContentTopLeft: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  rowContentTopRight: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'flex-start',
  },
  rowContentCenterLeft: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  rowContentBottomCenter: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
  columnContentTopCenter: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'center',
  },
  columnContentTopLeft: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  columnContentTopRight: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-end',
  },
  columnContentCenterRight: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
  columnContentCenterLeft: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  backgroundWhite: {
    backgroundColor: '#ffffff',
  },
  overlay: {
    backgroundColor: 'rgba(254, 222, 0, 0.3)',
    position: 'absolute',
    left: 0,
    top: 0,
    bottom: 0,
    right: 0,
  },
  spinner: {
    width: 50,
    height: 50,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-end',
    opacity: 1,
  },
  textCenter: {
    textAlign: 'center',
  },
  textSpinner: {
    fontSize: 20,
    color: Colors.CrossLightBlue,
    textAlign: 'center',
    fontWeight: 'normal',
    marginTop: 5,
    opacity: 1,
  },
  modal: {
    margin: 30,
    flex: 1,
    padding: 0,
    marginTop: 200,
  },
});

export const autoCompleteStyles = StyleSheet.create({
  // container: {
  //   backgroundColor: '#ffffff',
  // },
  padding: { padding: 20 },
  listView: {
    backgroundColor: '#ffffff',
  },
  textInputContainer: {
    borderBottomWidth: 0,
    borderTopWidth: 0,
  },
  textInput: {
    color: '#00539b',
    fontSize: 16,
  },
  description: {
    fontWeight: 'bold',
  },
  predefinedPlacesDescription: {
    color: '#fede00',
  },
});

export const sortableGridImagesStyles = StyleSheet.create({
  block: {
    flex: 1,
    margin: 8.5,
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center',
  },
  contentContainer: {
    flex: 9,
  },
  buttonView: {
    justifyContent: 'space-between',
    flexDirection: 'row',
    marginHorizontal: 15,
  },
  button: {
    width: 150,
    height: 50,
    alignItems: 'center',
    justifyContent: 'center',
  },
  imageButton: {
    flexDirection: 'row',
  },
  button1: {
    backgroundColor: '#257FE6',
  },
  button2: {
    backgroundColor: '#37AB33',
  },
  buttonText: {
    color: '#fff',
  },
  bottom: {
    width: screenWidth,
    alignSelf: 'center',
    flex: 1,
  },
  cam: {
    width: 16,
    height: 16,
    resizeMode: 'contain',
    marginRight: 2,
  },
});

export default styles;
