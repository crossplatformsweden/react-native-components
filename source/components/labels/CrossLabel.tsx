import React, { ReactNode } from 'react';
import {
  View,
  TextProps,
  Linking,
  TouchableHighlight,
  TextStyle,
} from 'react-native';
import {
  Text,
  Headline,
  Subheading,
  Title,
  Caption,
  Paragraph,
} from 'react-native-paper';
import styles from '../../styles';
import _ from 'lodash';

/**
 * Properties for the {@link CrossLabel} component. Control component type using {@link isTitle}, {@link isCaption},
 * {@link isHeadline}, {@link isSubheading} and {@link isParagraph}.
 *
 * Create an URL link using {@link onPressUrlTarget} and customize color using {@link linkColor}.
 *
 * Inherits react-native Text's {@link https://facebook.github.io/react-native/docs/text.html#props TextProps}
 */
export interface ICrossLabelProps extends TextProps {
  /**
   * An url to open when user clicks the label. Styles the label as a clickable "link"
   */
  onPressUrlTarget?: string;
  /**
   * Optional color for the link. Default is 'blue'. See also {@link onPressUrlTarget}
   */
  linkColor?: string | undefined;
  /**
   * Displays text as a {@link https://callstack.github.io/react-native-paper/headline.html react-native paper Headline component}
   */
  isHeadline?: boolean | undefined;
  /**
   * Displays text as a {@link https://callstack.github.io/react-native-paper/title.html react-native paper Title component}
   */
  isTitle?: boolean | undefined;
  /**
   * Displays text as a {@link https://callstack.github.io/react-native-paper/subheading.html react-native paper Subheading component}
   */
  isSubheading?: boolean | undefined;
  /**
   * Displays text as a {@link https://callstack.github.io/react-native-paper/paragraph.html react-native paper Paragraph component}
   */
  isParagraph?: boolean | undefined;
  /**
   * Displays text as a {@link https://callstack.github.io/react-native-paper/caption.html react-native paper Caption component}
   */
  isCaption?: boolean | undefined;
}

/**
 * Renders the relevant react-native-paper component based on props
 */
class TextComponent extends React.Component<ICrossLabelProps> {
  constructor(props: ICrossLabelProps) {
    super(props);
  }

  render() {
    let label: ReactNode = null;
    if (this.props.isHeadline) {
      label = <Headline {...this.props}>{this.props.children}</Headline>;
    }

    if (this.props.isSubheading) {
      label = <Subheading {...this.props}>{this.props.children}</Subheading>;
    }

    if (this.props.isTitle) {
      label = <Title {...this.props}>{this.props.children}</Title>;
    }

    if (this.props.isCaption) {
      label = <Caption {...this.props}>{this.props.children}</Caption>;
    }

    if (this.props.isParagraph) {
      label = <Paragraph {...this.props}>{this.props.children}</Paragraph>;
    }

    const url = _.get(this.props, ['onPressUrlTarget']);

    if (url && !label)
      return (
        <Text {...this.props} onPress={() => Linking.openURL(url.toString())}>
          {this.props.children}
        </Text>
      );

    if (url && label)
      return (
        <TouchableHighlight onPress={() => Linking.openURL(url.toString())}>
          {label}
        </TouchableHighlight>
      );

    return label ? (
      { ...label }
    ) : (
      <Text {...this.props}>{this.props.children}</Text>
    );
  }
}

/**
 * Wrapper for {@link https://callstack.github.io/react-native-paper/text.html react-native-paper Text} component
 * that displays as a clickable link if {@link ICrossLabelProps.onPressUrlTarget} is supplied.
 *
 * Link color can also be specified through {@link ICrossLabelProps.linkColor}.
 *
 * Properties are {@link ICrossLabelProps}
 *
 * @param param0 {@link ICrossLabelProps}
 * @returns {@link View}
 */
export class CrossLabel extends React.Component<ICrossLabelProps> {
  constructor(props: ICrossLabelProps) {
    super(props);
  }

  render() {
    return (
      <View style={styles.container}>
        <TextComponent
          {...this.props}
          style={[
            this.props.onPressUrlTarget ? linkStyle : {},
            this.props.style,
            this.props.linkColor ? { color: this.props.linkColor } : {},
          ]}
        >
          {this.props.children}
        </TextComponent>
      </View>
    );
  }
}

const linkStyle: TextStyle = {
  color: 'blue',
  fontWeight: 'bold',
  backgroundColor: 'transparent',
};

export default CrossLabel;
