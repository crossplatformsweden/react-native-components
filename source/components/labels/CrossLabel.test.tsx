import React from 'react';
import CrossLabel, { ICrossLabelProps } from './CrossLabel';
import TestRenderer from 'react-test-renderer';
import {
  Headline,
  Title,
  Subheading,
  Caption,
  Paragraph,
  Text,
} from 'react-native-paper';
import { Colors } from '../../styles';
import { Linking } from 'react-native';

jest.unmock('react-native');
jest.unmock('./CrossLabel');

describe('components', () => {
  describe('<CrossLabel />', () => {
    describe('`ICrossLabelProps` interface', () => {
      const propsMock: ICrossLabelProps = {};
      it('Should match snapshot', () => {
        expect(propsMock).toMatchSnapshot();
      });

      it('Should have `isTitle` property that is undefined', () => {
        expect(propsMock.isTitle).toBeUndefined();
      });
    });

    it('Component should render', () => {
      const wrapper = TestRenderer.create(<CrossLabel />);
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('When `isHeadline` is `true` should contain <Headline />', () => {
      const wrapper = TestRenderer.create(
        <CrossLabel isHeadline={true}>
          Crossplatform (isHeadline=true)
        </CrossLabel>
      );
      const child = wrapper.root.findByType(Headline);
      expect(child.type).toEqual('Text');
    });

    it('When `isTitle` is `true` should contain <Headline />', () => {
      const wrapper = TestRenderer.create(
        <CrossLabel isTitle={true}>&lt;CrossLabel isTitle=true&gt;</CrossLabel>
      );
      const child = wrapper.root.findByType(Title);
      expect(child.type).toEqual('Text');
    });

    it('When `isSubheading` is `true` should contain <Subheading />', () => {
      const wrapper = TestRenderer.create(
        <CrossLabel isSubheading={true} style={{ marginTop: 5 }}>
          isSubHeading = true
        </CrossLabel>
      );
      const child = wrapper.root.findByType(Subheading);
      expect(child.type).toEqual('Text');
    });

    it('When `isCaption` is `true` should contain <Caption />', () => {
      const wrapper = TestRenderer.create(
        <CrossLabel
          isCaption={true}
          style={{ color: Colors.CrossLightBlue, marginTop: 10 }}
        >
          isCaption=true, custom color
        </CrossLabel>
      );
      const child = wrapper.root.findByType(Caption);
      expect(child.type).toEqual('Text');
    });

    it('When `isParagraph` is `true` should contain <Paragraph />', () => {
      const wrapper = TestRenderer.create(
        <CrossLabel isParagraph={true} style={{ marginTop: 5 }}>
          isParagraph = true
        </CrossLabel>
      );
      const child = wrapper.root.findByType(Paragraph);
      expect(child.type).toEqual('Text');
    });

    it('When `onPressUrlTarget` is url should be clickable', () => {
      const wrapper = TestRenderer.create(
        <CrossLabel
          onPressUrlTarget='https://www.typescriptlang.org/'
          style={{ marginTop: 20, marginBottom: 10 }}
        >
          Clickable link (onPressUrlTarget):
        </CrossLabel>
      );
      const child = wrapper.root.findByType(Text);
      child.props.onPress();
      expect(Linking.openURL).toBeCalledWith('https://www.typescriptlang.org/');
    });
  });
});
