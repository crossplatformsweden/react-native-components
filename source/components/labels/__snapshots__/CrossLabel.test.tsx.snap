// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <CrossLabel /> \`ICrossLabelProps\` interface Should match snapshot 1`] = `Object {}`;

exports[`components <CrossLabel /> Component should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <Text
    style={
      Array [
        Object {},
        undefined,
        Object {},
      ]
    }
  />
</View>
`;
